File: common/hogvm/python/stl/crypto.py
import hmac
import base64
import hashlib
from typing import Literal


def encode_digest(encoding: Literal["hex", "base64", "base64url", "binary"], digest: bytes) -> str:
    if encoding == "hex":
        return digest.hex()
    elif encoding == "base64":
        return base64.b64encode(digest).decode()
    elif encoding == "base64url":
        return base64.urlsafe_b64encode(digest).decode().rstrip("=")
    elif encoding == "binary":
        return digest.decode("latin1")


def md5(data: str | None, encoding: Literal["hex", "base64", "base64url", "binary"] = "hex") -> str | None:
    if data is None:
        return None
    # nosemgrep: python.lang.security.insecure-hash-algorithms-md5.insecure-hash-algorithm-md5
    digest = hashlib.md5(data.encode()).digest()

    return encode_digest(encoding, digest)


def sha256(data: str | None, encoding: Literal["hex", "base64", "base64url", "binary"] = "hex") -> str | None:
    if data is None:
        return None
    digest = hashlib.sha256(data.encode()).digest()

    return encode_digest(encoding, digest)


def sha1(data: str | None, encoding: Literal["hex", "base64", "base64url", "binary"] = "hex") -> str | None:
    if data is None:
        return None
    # SHA-1 is only here because vendors sign webhooks with it. Never pick it for new work.
    # nosemgrep: python.lang.security.insecure-hash-algorithms.insecure-hash-algorithm-sha1
    digest = hashlib.sha1(data.encode()).digest()

    return encode_digest(encoding, digest)


def _hmac_chain(data: list, digestmod: str, encoding: Literal["hex", "base64", "base64url", "binary"]) -> str:
    if len(data) < 2:
        raise ValueError("Data array must contain at least two elements.")

    hmac_obj = hmac.new(data[0].encode(), data[1].encode(), digestmod)
    for i in range(2, len(data)):
        hmac_obj = hmac.new(hmac_obj.digest(), data[i].encode(), digestmod)

    digest = hmac_obj.digest()

    return encode_digest(encoding, digest)


def sha256HmacChain(data: list, encoding: Literal["hex", "base64", "base64url", "binary"] = "hex") -> str:
    return _hmac_chain(data, "sha256", encoding)


def sha1HmacChain(data: list, encoding: Literal["hex", "base64", "base64url", "binary"] = "hex") -> str:
    return _hmac_chain(data, "sha1", encoding)

File: common/hogvm/typescript/src/stl/crypto.ts
import { ExecOptions } from '../types'

export function sha256(
    data: string | null,
    encoding: 'hex' | 'base64' | 'base64url' | 'binary' = 'hex',
    options?: ExecOptions
): string | null {
    if (data === null) {
        return null
    }
    const crypto = options?.external?.crypto
    if (!crypto) {
        throw new Error('The crypto module is required for "sha256Hex" to work.')
    }
    return crypto.createHash('sha256').update(data).digest(encoding)
}

export function md5(
    data: string | null,
    encoding: 'hex' | 'base64' | 'base64url' | 'binary' = 'hex',
    options?: ExecOptions
): string | null {
    if (data === null) {
        return null
    }
    const crypto = options?.external?.crypto
    if (!crypto) {
        throw new Error('The crypto module is required for "md5Hex" to work.')
    }
    return crypto.createHash('md5').update(data).digest(encoding)
}

// SHA-1 is only here because vendors sign webhooks with it. Never pick it for new work.
export function sha1(
    data: string | null,
    encoding: 'hex' | 'base64' | 'base64url' | 'binary' = 'hex',
    options?: ExecOptions
): string | null {
    if (data === null) {
        return null
    }
    const crypto = options?.external?.crypto
    if (!crypto) {
        throw new Error('The crypto module is required for "sha1Hex" to work.')
    }
    return crypto.createHash('sha1').update(data).digest(encoding)
}

function hmacChain(
    algorithm: 'sha256' | 'sha1',
    name: string,
    data: string[],
    encoding: 'hex' | 'base64' | 'base64url' | 'binary',
    options?: ExecOptions
): string {
    const crypto = options?.external?.crypto
    if (!crypto) {
        throw new Error(`The crypto module is required for "${name}" to work.`)
    }
    if (data.length < 2) {
        throw new Error('Data array must contain at least two elements.')
    }
    let hmac = crypto.createHmac(algorithm, data[0])
    hmac.update(data[1])
    for (let i = 2; i < data.length; i++) {
        hmac = crypto.createHmac(algorithm, hmac.digest())
        hmac.update(data[i])
    }
    return hmac.digest(encoding)
}

export function sha256HmacChain(
    data: string[],
    encoding: 'hex' | 'base64' | 'base64url' | 'binary' = 'hex',
    options?: ExecOptions
): string {
    return hmacChain('sha256', 'sha256HmacChainHex', data, encoding, options)
}

export function sha1HmacChain(
    data: string[],
    encoding: 'hex' | 'base64' | 'base64url' | 'binary' = 'hex',
    options?: ExecOptions
): string {
    return hmacChain('sha1', 'sha1HmacChainHex', data, encoding, options)
}
